### PR TITLE
Initial docs draft: Overview, History Service, Workflow Lifecycle

### PR DESCRIPTION
**Pull request by dandavison**
**What changed?**
- Initial draft of system architecture / internal implementation docs for
  - History Service
  - Temporal overview

Links to rendered Markdown for review:

- [Temporal overview](https://github.com/dandavison/temporalio-temporal/tree/docs-history-service/docs)
- [History Service](https://github.com/dandavison/temporalio-temporal/blob/docs-history-service/docs/history-service.md)
- [Workflow Lifecycle](https://github.com/dandavison/temporalio-temporal/blob/docs-history-service/docs/workflow-lifecycle.md) [based on @yiminc's diagrams]

**Is hotfix candidate?**
- No

### CODE REVIEW

File: docs/README.md
Comment (alexshtin) [inline docs/README.md:8]: I would split "internal services" and "overall concepts". Especially in this list, it sounds like "Temporal Workers" section is about "service/worker" but it is not. Let's add a section with "Internal services" and list all 4 there: Frontend, History, Matching, Intreal worker. Link FE and IW to some "./TODO.md".  And then "Concepts" (or whatever section) with two other items.

File: docs/matching-service.md
Comment (antlai-temporal) [inline docs/matching-service.md:8]: respnsible typo

File: docs/history-service.md
Comment (antlai-temporal) [inline docs/history-service.md:230]: "event" is overloaded, maybe "inputs" to ensure less confusion with history events

File: docs/history-service.md
Comment (antlai-temporal) [inline docs/history-service.md:273]: I guess you will need to flush the log of history events before doing the transaction, otherwise you could have an id in mutable state for an event that you cannot find anywhere... Or is the whole event also stored in mutable state?

File: docs/history-service.md
Comment (antlai-temporal) [inline docs/history-service.md:274]: The "Transfer Task queue processor" does not appear in the diagram. A comment of how it deals with duplicates, and how it records progress, may be useful...

File: docs/history-service.md
Comment (alexshtin) [inline docs/history-service.md:76]: Why referencing specific commit, not `main` branch? Yes, it can be broken, but link to inactive code is even worse.

Comment (alexshtin) [reply docs/history-service.md:76]: @flossypurse has a tool to manage links to the specific line of code in refactor proof way.

File: docs/history-service.md
Comment (alexshtin) [inline docs/history-service.md:166]: :+1:

Comment (alexshtin) [reply docs/history-service.md:166]: may be add some **bold** here.

File: docs/history-service.md
Comment (alexshtin) [inline docs/history-service.md:206]: So, TT is not only for RPC (for example `CloseExecutionTask`). I don't think that we need to list all cases here. These two, you mentioned, are most important, for sure. But I think it worth to say that there are other cases.

File: docs/history-service.md
Comment (alexshtin) [inline docs/history-service.md:212]: All timeouts have corresponding timer task.

File: docs/history-service.md
Comment (alexshtin) [inline docs/history-service.md:272]: It worth to explain/point out (may be not here) why we store 6 different objects in `executions` table.

File: docs/history-service.md
Comment (alexshtin) [inline docs/history-service.md:274]: Make "Transactional Outbox" a link to https://microservices.io/patterns/data/transactional-outbox.html ?

File: README.md
Comment (alexshtin) [inline README.md:24]: There is a link to `CONTRIBUTING.md` on line 62. I think it should go to `docs` dir too.

File: README.md
Comment (alexshtin) [inline README.md:11]: Finally! It is :man_facepalming: that we are still "microservice orchestration platform" here.

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:76]: > Why referencing specific commit, not `main` branch? Yes, it can be broken, but link to inactive code is even worse.

Links to `main` branch rapidly end up pointing at nonsense. The permalinks I'm using will always point to what they are intended to point to. So, it is guaranteed that the reader when following the link will get to the function, and it will be in a context where it made sense. So IMO if we include links with line numbers they should either be permalinks, or CI would need to guarantee validity of the link to `main`.

I'd still like to consider the possibility that we do include links with line numbers; I am finding them very useful. If it has moved on main then the reader will have a function name that they can search for in their IDE. But we would ideally like to detect when it has moved on main and when it has disappeared from main. This could in principle be done in CI.

File: docs/history-service.md
Comment (dnr) [inline docs/history-service.md:64]:
```suggestion
The History Service has many other responsibilities which are not covered here, for example: Shard Membership and Ownership, Replication and conflict resolution, Visibility, Archival, Workflow Reset.
```

history doesn't really do anything for versioning besides passing around some metadata. and it doesn't do anything special for scheduled workflows either (besides just that the scheduler workflow is a workflow, but it's not treated specially by history)

File: docs/history-service.md
Comment (dnr) [inline docs/history-service.md:89]: there are a few things in mutable state that are not derived from events. iirc, activity heartbeats are one. not sure if there are any other significant ones.

File: docs/history-service.md
Comment (dnr) [inline docs/history-service.md:195]: those are only two types of transfer task. it looks like there are six more (close, cancel, signal, start child, auto-reset, delete).

File: docs/history-service.md
Comment (dnr) [inline docs/history-service.md:235]: what about start child/signal external from another workflow? you could say they're indirectly from an rpc from a worker (they do come from workflow task commands), but it's indirect, the rpc comes from another history shard's transfer queue. so maybe that's a fourth type.

File: docs/history-service.md
Comment (dnr) [inline docs/history-service.md:274]: it might be good to talk about the other parts of the history↔matching protocol here. i.e. if matching can't dispatch a task immediately (sync), it has to store it durably in its own persistence before returning from the add task rpc. and matching has to get a successful response from a recordxtaskstarted rpc to history before dispatching a task to a worker.

File: docs/history-service.md
Comment (dnr) [inline docs/history-service.md:90]: maybe say somewhere that some number of mutable state are cached in memory?

File: docs/workflow-lifecycle.md
Comment (josh-berry) [inline docs/workflow-lifecycle.md:85]: A little more context here might be helpful—why would the worker send `ScheduleActivityTask`? For example:

```suggestion
3. **The workflow launches an activity, causing the Worker to send a `ScheduleActivityTask` command back to the Frontend**
```

File: docs/workflow-lifecycle.md
Comment (josh-berry) [inline docs/workflow-lifecycle.md:116]: Similarly, we're making an assumption here that the activity is running on the same worker as the workflow (which is a good assumption to make, but we might want to mention it here):

```suggestion
**4. The Worker dequeues the Activity task and executes the activity.**
  - (Note: This Worker may or may not be the same worker that launched the activity, depending on what other workers are available, the type of the activity, etc.)
```

File: docs/workflow-lifecycle.md
Comment (josh-berry) [inline docs/workflow-lifecycle.md:145]:
```suggestion
5. **Once the activity is completed, the Worker running the Activity sends `RespondActivityTaskCompleted`, containing the activity result**
```

File: docs/workflow-lifecycle.md
Comment (josh-berry) [inline docs/workflow-lifecycle.md:176]: I know this is perhaps oversimplified (since we're not talking about replay), but:

```suggestion
6. **The Worker running the workflow dequeues the Workflow Task**
```

File: docs/workflow-lifecycle.md
Comment (josh-berry) [inline docs/workflow-lifecycle.md:187]: The workflow is complete at this point, yeah?

File: docs/workflow-lifecycle.md
Comment (josh-berry) [inline docs/workflow-lifecycle.md:216]: Should we also talk about workflow replay in this doc?

**Review by josh-berry — COMMENTED** [review]
Nothing in my review comments should hold up the bus. :)

File: README.md
Comment (tdeebswihart) [reply README.md:24]: FYI @dandavison Alex recently rediscovered our existing docs folder: at `develop/docs`

File: docs/README.md
Comment (dandavison) [reply docs/README.md:8]: Thanks, changed to:

<table>
  <tr>
    <td>image
    </td>
  </tr
</table>

File: docs/README.md
Comment (dandavison) [reply docs/README.md:16]: Right! Fixed.

File: docs/README.md
Comment (dandavison) [reply docs/README.md:26]: Thanks, I've taken that verbatim.

File: docs/README.md
Comment (dandavison) [reply docs/README.md:41]: Thanks, fixed.

File: docs/README.md
Comment (dandavison) [reply docs/README.md:51]: Thanks, added

> In addition, Queries are implemented via Query Tasks, which are similar to Workflow Tasks but result only in Query results being sent to the server (do not cause the Workflow to advance).

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:39]: Good call, done.

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:42]: I think it would be hard to do that without damaging the clarity of this diagram. Every arrow on the right is labeled `enqueue XXX Task` and I'm hoping that that the repeated word "enqueue" will be enough to cause people to realize what is doing the dequeuing. The full cycle is depicted in the diagram at the top of the page; but this page is really documenting the History Service so I think it makes sense here to show the History Service acting as server and client, but not to show the parts of the cycle that don't involve History Service.

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:89]: Thanks, changed to "most of this data could...".

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:90]: Thanks, added.

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:166]: Added italics.

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:195]: Thanks for pointing this out @dnr and @alexshtin. I think I had not picked up on that. I've added
> (In addition to enqueuing Workflow and Activity tasks, there are several other types of Transfer Task.)

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:212]: Right, it does say "For example:", and I wanted to give a concrete example here.

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:230]: Good call, changed to "inputs".

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:235]: Thanks, added

> - A different Workflow Execution performing start-child-workflow / signal-external-workflow

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:274]: Agreed, that seems like a reasonable source to link to for our usage. Done.

Comment (dandavison) [reply docs/history-service.md:274]: I think I'd like to leave that to a subsequent update of this document, or for the (currently missing) `docs/matching-service.md`, and quite possibly by a different author. (I'm already at risk of over-reaching myself here.)

File: README.md
Comment (dandavison) [reply README.md:24]: > There is a link to CONTRIBUTING.md on line 62. I think it should go to docs dir too.
Thanks, added.

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md]: That would be great. My aim here was to create an initial foundation, that makes it much easier for other people to document specific areas. I think I'm already at risk of over-reaching myself -- maybe @yycptt or someone else would like to take a stab at that in a follow-on PR.

File: docs/matching-service.md
Comment (dandavison) [reply docs/matching-service.md:6]: Thanks, changed to

> A single Task Queue is responsible for delivering tasks relating to many Workflow Executions.

File: docs/workflow-lifecycle.md
Comment (dandavison) [reply docs/workflow-lifecycle.md:116]: Good call. I've changed it to "A Worker".

File: docs/workflow-lifecycle.md
Comment (dandavison) [reply docs/workflow-lifecycle.md:176]: Thanks, changed to "A worker dequeues"

File: docs/workflow-lifecycle.md
Comment (dandavison) [reply docs/workflow-lifecycle.md:187]: Yes, good point, I've changed it to make it clearer how the workflow execution is completed:

image

File: docs/workflow-lifecycle.md
Comment (dandavison) [reply docs/workflow-lifecycle.md:216]: I think that would be good. @Sushisource has some sequence diagrams which it would be good to integrate into this page. Their center of gravity is slightly more SDK, but they'd make sense together. (My aim here is to provide a foundation with basics that makes it easy for others to really improve the docs.)

File: docs/history-service.md
Comment (MichaelSnowden) [inline docs/history-service.md:95]: I think the actual reason we do this is because of Cassandra. We have some relational data, like the pending activites for a workflow, and we want to do 2 relational DB-esque things that Cassandra doesn't really support:
1. Update this relational data (e.g. add a pending activity) transactionally
2. Query this data in an indexed fashion (e.g. get all child workflows of a workflow)

In a normal RDBMS, you'd store this data in separate tables, with a logical data model (e.g. an activities table with a foreign key on the workflow). However, these things don't really exist, and there's no working analog in Cassandra:
1. You can't update data in different tables in a single transaction in Cassandra, so we keep almost everything in one polymorphic "executions" table
2. Even if we didn't need transactional updates, we wouldn't be able to query this data efficiently because adding secondary indices in Cassandra has so many drawbacks, that we just almost never do it. Instead, we configure it so that all data you need is queryable via <full partition key> + <some prefix of the clustering key>, or it's colocated in the same row.

So, you're right that we keep track of some state, which could be inferred from just events, for speed. However, I think it's also worth mentioning that the reason all of this tate is thrown together in one giant "mutable state" object instead of a more traditional relational data model is that we need to support non-relational databases like Cassandra.

This particular distinction isn't very important, but I think these limitations explain a lot of our unintuitive design choices, so they're worth pointing out.

**Review by MichaelSnowden — APPROVED** [review]

File: docs/history-service.md
Comment (dandavison) [reply docs/history-service.md:82]: I'm going to merge this with the permalinks, but I agree it's not clear what the best way to do this is, so we can settle on our preferred approach and change if that's what we decide. Some options are

- permalinks with line numbers
- link to file only (in `main`), mentioning function/struct name
- no code links at all

In principle we could even use a format that permits a CI check to be written that confirms that the referenced name occurs on the referenced line. Or it might even be possible to automatically compute updated permalinks with sufficient accuracy, again in principle, not a high priority project.

**Comment by dandavison** [comment]
Merged, thanks for all the reviewing everyone.

**Pull request merged**
